bgpv2: Cleanup BGPInstance reconciler metadata

rastislavs opened the pull request:
Replace global BGPv2 reconciler metadata kept in `BGPInstance` with metadata kept privately in individual reconcilers.

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -66,14 +68,11 @@ type PeerData struct {
 
 // NeighborReconcilerMetadata keeps a map of running peers to peer configuration.
-// key is the peer name.
+// Key is the instance name.
```
@harsimran-pabla do you remember why `NeighborReconcilerMetadata` is `map[string][]*PeerData` keyed by the instance name, even though metadata is anyways instance-specific? To me it seems it could be just `[]*PeerData`

harsimran-pabla replied:
Hmm, strange I think I might have made a mistake. Intention was probably to store map of PeerData, with key as neighbor ID. Upsert metadata would be simple map update instead of iterating each neighbor.

rastislavs replied:
OK, I see. I pushed an additional commit which cleans this up (https://github.com/cilium/cilium/pull/34426/commits/11dbba0baa4c33b92a447102127f3ff425e1e21e)

harsimran-pabla commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -27,6 +27,7 @@ type NeighborReconciler struct {
 	SecretStore  store.BGPCPResourceStore[*slim_corev1.Secret]
 	PeerConfig   store.BGPCPResourceStore[*v2alpha1.CiliumBGPPeerConfig]
 	DaemonConfig *option.DaemonConfig
+	metadata     map[uint32]NeighborReconcilerMetadata
```
General comment around ASN as key for metadata, should it be instance name instead?

rastislavs replied:
This was addressed in a separate PR: https://github.com/cilium/cilium/pull/35207

harsimran-pabla commented on pkg/bgpv1/manager/instance/fake_instance.go:
```diff
@@ -10,8 +10,8 @@ import (
-		Metadata: make(map[string]any),
+		Config: nil,
+		Router: types.NewFakeRouter(),
+		Global: types.BGPGlobal{},
```
I suggest having separate commit for introducing Global config, would be easier to review what it impacts.

rastislavs replied:
split the change into 2 commits

rastislavs committed:
bgpv2: Keep reconciler metadata private within reconciler

Before this change, all BGP reconcilers were storing their instance-specific
state (metadata) in a common untyped BGP instance map. This was due to
historical reasons, as in the past we did not have the Init(instance)
and Cleanup(instance) reconciler methods. As we now have them, we can
store all instance-specific reconciler state privately within the reconciler.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
bgpv2: Remove Metadata and ASN fields from BGPInstance

Since the global instance Metadata is no longer used in the
reconcilers, remove it from the BGPInstance struct.
Also remove the ASN field, as it can be found in Global.ASN,
and is no longer that frequently accessed from the reconcilers.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
bgpv2: Simplify NeighborReconcilerMetadata structure

There is no need to have PeerData keyed by instance name
in the NeighborReconcilerMetadata, as this metadata is
anyways alwyas instance-specific (and the map would always
contain just single instance).
Instead, we store peer data in a map keyed by peer name,
which makes peer data access simpler - without the need
for slice iteration and manipulation.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

harsimran-pabla reviewed (approved):
Nice cleanup ❤️

YutaroHayakawa reviewed (approved):
👍 👍 👍
